File: apps/mobile/android/app/src/main/java/dev/cm4ker/meshnet/MeshTcpPlugin.java
package dev.cm4ker.meshnet;

import android.util.Base64;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * A TCP link to a radio on the network: {@code window.Capacitor.Plugins.MeshTcp},
 * the same plugin the iOS app carries (MeshTcpPlugin.swift), used by the web
 * client's {@code transports/capacitorTcp.ts}. Companion firmware built with Wi-Fi
 * listens on a port (5000 unless built otherwise) and speaks the framed stream
 * of its USB serial; the framing is the client's, this only moves bytes.
 *
 * <p>{@code open({host, port, timeout})} answers {@code {id}}; {@code write({id, data})}
 * takes base64; {@code close({id})}. Events: {@code data {id, data}} as bytes arrive, and
 * {@code closed {id, error}} when a connection ends without being closed from here.
 */
@CapacitorPlugin(name = "MeshTcp")
public class MeshTcpPlugin extends Plugin {
    private final Map<String, Socket> sockets = new ConcurrentHashMap<>();
    /** Connects and reads, one thread per connection. */
    private final ExecutorService readers = Executors.newCachedThreadPool();
    /** Writes, one at a time, so frames go out in the order they were sent. */
    private final ExecutorService writer = Executors.newSingleThreadExecutor();

    @PluginMethod
    public void open(PluginCall call) {
        String host = call.getString("host", "").trim();
        int port = call.getInt("port", 5000);
        double timeout = call.getDouble("timeout", 10.0);
        if (host.isEmpty()) {
            call.reject("an address is needed");
            return;
        }
        if (port <= 0 || port >= 65536) {
            call.reject(port + " is not a port");
            return;
        }
        String address = host + ":" + port;
        readers.execute(() -> {
            Socket socket = new Socket();
            try {
                socket.setTcpNoDelay(true);
                socket.setKeepAlive(true);
                socket.connect(new InetSocketAddress(host, port), (int) (timeout * 1000));
            } catch (IOException e) {
                closeQuietly(socket);
                call.reject(address + " did not answer: " + e.getMessage());
                return;
            }
            String id = UUID.randomUUID().toString();
            sockets.put(id, socket);
            JSObject result = new JSObject();
            result.put("id", id);
            call.resolve(result);
            read(id, socket);
        });
    }

    @PluginMethod
    public void write(PluginCall call) {
        String id = call.getString("id", "");
        String data = call.getString("data");
        if (data == null) {
            call.reject("write needs base64 data");
            return;
        }
        byte[] bytes = Base64.decode(data, Base64.DEFAULT);
        writer.execute(() -> {
            Socket socket = sockets.get(id);
            if (socket == null) {
                call.reject("not connected");
                return;
            }
            try {
                OutputStream out = socket.getOutputStream();
                out.write(bytes);
                out.flush();
                call.resolve();
            } catch (IOException e) {
                call.reject(e.getMessage());
            }
        });
    }

    @PluginMethod
    public void close(PluginCall call) {
        Socket socket = sockets.remove(call.getString("id", ""));
        if (socket != null) closeQuietly(socket);
        call.resolve();
    }

    @Override
    protected void handleOnDestroy() {
        for (Socket socket : sockets.values()) closeQuietly(socket);
        sockets.clear();
        readers.shutdownNow();
        writer.shutdownNow();
    }

    private void read(String id, Socket socket) {
        byte[] buffer = new byte[4096];
        String reason = "the radio closed the connection";
        try {
            InputStream in = socket.getInputStream();
            int n;
            while ((n = in.read(buffer)) > 0) {
                JSObject event = new JSObject();
                event.put("id", id);
                event.put("data", Base64.encodeToString(buffer, 0, n, Base64.NO_WRAP));
                notifyListeners("data", event);
            }
        } catch (IOException e) {
            reason = e.getMessage();
        }
        // A socket closed from here is already out of the table, and says nothing.
        if (sockets.remove(id) == null) return;
        closeQuietly(socket);
        JSObject event = new JSObject();
        event.put("id", id);
        event.put("error", reason);
        notifyListeners("closed", event);
    }

    private static void closeQuietly(Socket socket) {
        try {
            socket.close();
        } catch (IOException ignored) {
            // Already closed.
        }
    }
}

File: apps/mobile/ios/App/App/MeshTcpPlugin.swift
import Capacitor
import Foundation
import Network

/// A TCP link to a radio on the network: `window.Capacitor.Plugins.MeshTcp`, used
/// by the web client's `transports/capacitorTcp.ts`. Companion firmware built with
/// Wi-Fi listens on a port (5000 unless built otherwise) and speaks the same framed
/// stream as its USB serial; the framing is the client's, this only moves bytes.
///
/// - `open({ host, port, timeout })`: answers `{ id }` once the connection is up.
/// - `write({ id, data })`: `data` is base64.
/// - `close({ id })`.
/// - Events: `data` `{ id, data }` (base64) as bytes arrive, and `closed`
///   `{ id, error? }` when the radio or the network ends a connection that was
///   not closed from here.
///
/// Every connection lives on one serial queue, which is also where the table of
/// them is read and written.
@objc(MeshTcpPlugin)
final class MeshTcpPlugin: CAPPlugin, CAPBridgedPlugin {
    let identifier = "MeshTcpPlugin"
    let jsName = "MeshTcp"
    let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "write", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "close", returnType: CAPPluginReturnPromise),
    ]

    private let queue = DispatchQueue(label: "dev.cm4ker.meshnet.tcp")
    private var connections: [String: NWConnection] = [:]

    @objc func open(_ call: CAPPluginCall) {
        guard let host = call.getString("host")?.trimmingCharacters(in: .whitespaces), !host.isEmpty else {
            call.reject("an address is needed")
            return
        }
        let number = call.getInt("port") ?? 5000
        guard number > 0, number < 65536, let port = NWEndpoint.Port(rawValue: UInt16(number)) else {
            call.reject("\(number) is not a port")
            return
        }
        let timeout = call.getDouble("timeout") ?? 10

        let tcp = NWProtocolTCP.Options()
        tcp.noDelay = true
        tcp.enableKeepalive = true
        tcp.keepaliveIdle = 15
        let connection = NWConnection(host: NWEndpoint.Host(host), port: port, using: NWParameters(tls: nil, tcp: tcp))
        let id = UUID().uuidString
        let address = "\(host):\(number)"

        // A connection that cannot be made yet waits rather than fails: the
        // network may come back, or the phone may be showing its Local Network
        // prompt. So it is given until the timeout, and a wait's reason is kept
        // to say why it gave up.
        var settled = false
        var waitingFor: String?
        let settle: (String?) -> Void = { [weak self] error in
            guard let self, !settled else { return }
            settled = true
            if let error {
                connection.cancel()
                call.reject(error)
                return
            }
            self.connections[id] = connection
            call.resolve(["id": id])
            self.receive(id, connection)
        }

        connection.stateUpdateHandler = { [weak self] state in
            switch state {
            case .ready:
                settle(nil)
            case .waiting(let error):
                waitingFor = error.localizedDescription
            case .failed(let error):
                if settled {
                    self?.dropped(id, error.localizedDescription)
                    connection.cancel()
                } else {
                    settle("\(address): \(error.localizedDescription)")
                }
            default:
                break
            }
        }
        queue.asyncAfter(deadline: .now() + timeout) {
            settle("\(address) did not answer" + (waitingFor.map { ": \($0)" } ?? ""))
        }
        connection.start(queue: queue)
    }

    @objc func write(_ call: CAPPluginCall) {
        guard let id = call.getString("id"), let data = call.getString("data").flatMap({ Data(base64Encoded: $0) }) else {
            call.reject("write needs an id and base64 data")
            return
        }
        queue.async {
            guard let connection = self.connections[id] else {
                call.reject("not connected")
                return
            }
            connection.send(content: data, completion: .contentProcessed { error in
                if let error {
                    call.reject(error.localizedDescription)
                } else {
                    call.resolve()
                }
            })
        }
    }

    @objc func close(_ call: CAPPluginCall) {
        let id = call.getString("id") ?? ""
        queue.async {
            self.connections.removeValue(forKey: id)?.cancel()
            call.resolve()
        }
    }

    private func receive(_ id: String, _ connection: NWConnection) {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 4096) { [weak self] data, _, isComplete, error in
            guard let self else { return }
            if let data, !data.isEmpty {
                self.notifyListeners("data", data: ["id": id, "data": data.base64EncodedString()])
            }
            if let error {
                self.dropped(id, error.localizedDescription)
                connection.cancel()
            } else if isComplete {
                self.dropped(id, "the radio closed the connection")
                connection.cancel()
            } else {
                self.receive(id, connection)
            }
        }
    }

    /// Tells the page about a connection that ended without its asking. One it
    /// closed itself is already out of the table and says nothing.
    private func dropped(_ id: String, _ reason: String) {
        guard connections.removeValue(forKey: id) != nil else { return }
        notifyListeners("closed", data: ["id": id, "error": reason])
    }
}
